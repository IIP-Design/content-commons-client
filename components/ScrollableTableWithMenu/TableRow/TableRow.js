import React, { Fragment, useState } from 'react';
import PropTypes from 'prop-types';
import dynamic from 'next/dynamic';
import { formatDate } from 'lib/utils';
import { Table } from 'semantic-ui-react';
import TableMobileDataToggleIcon from 'components/ScrollableTableWithMenu/TableMobileDataToggleIcon/TableMobileDataToggleIcon';

const MyProjectPrimaryCol = dynamic( () => import(
  /* webpackChunkName: "myProjectPrimaryCol" */
  'components/admin/Dashboard/MyProjects/MyProjectPrimaryCol/MyProjectPrimaryCol'
) );

const TeamProjectPrimaryCol = dynamic( () => import(
  /* webpackChunkName: "teamProjectPrimaryCol" */
  'components/admin/Dashboard/TeamProjects/TeamProjectPrimaryCol/TeamProjectPrimaryCol'
) );

const TableRow = ( { d, tableHeaders, projectTab } ) => {
  const [displayMobileData, setDisplayMobileData] = useState( false );

  const handleDisplayMobileData = () => {
    setDisplayMobileData( prevDisplayMobileData => !prevDisplayMobileData );
  };

  if ( !d ) return null;

  return (
    <Table.Row
      className={ displayMobileData ? 'activeTableRow' : '' }
    >
      { tableHeaders.map( ( header, i ) => (
        <Table.Cell
          data-header={ header.label }
          key={ `${d.id}_${header.name}` }
          className={ `items_table_item${d.status === 'PUBLISHING' ? ' isPublishing' : ''}` }
        >
          { i === 0
            ? (
              // Table must include .primary_col div for fixed column
              <Fragment>
                <div className="primary_col">
                  { projectTab === 'myProjects' && (
                    <MyProjectPrimaryCol
                      d={ d }
                      header={ header }
                    />
                  ) }
                  { projectTab === 'teamProjects' && (
                    <TeamProjectPrimaryCol
                      d={ d }
                      header={ header }
                    />
                  ) }
                </div>
                <TableMobileDataToggleIcon
                  isOpen={ displayMobileData }
                  toggleDisplay={ handleDisplayMobileData }
                />
              </Fragment>
            )
            : (
              <Fragment>
                <span>
                  <div className="items_table_mobileHeader">{ header.label }</div>
                  {
                    header.name === 'createdAt' || header.name === 'updatedAt'
                      ? formatDate( d[header.name] )
                      : d[header.name]
                  }
                </span>
                <br />
                { header.label === 'CREATED'
                  ? <span>{ d.status }</span>
                  : null }
              </Fragment>
            ) }
        </Table.Cell>
      ) ) }
    </Table.Row>
  );
};

TableRow.propTypes = {
  d: PropTypes.object,
  tableHeaders: PropTypes.array,
  projectTab: PropTypes.string
};

export default TableRow;
