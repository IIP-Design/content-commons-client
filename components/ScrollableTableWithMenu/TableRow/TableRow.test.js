import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import TableRow from './TableRow';

/**
 * Need to mock Next.js dynamic imports
 * in order for this test suite to run.
 */
jest.mock( 'next/dynamic', () => () => 'VideoDetailsPopup' );

const myProjectsProps = {
  d: {
    id: 'd137',
    createdAt: '2019-05-09T18:33:03.368Z',
    updatedAt: '2019-05-09T18:33:03.368Z',
    team: 'IIP Video Production',
    author: 'Jane Doe',
    projectTitle: 'Test Title 1',
    status: 'PUBLISHED',
    visibility: 'INTERNAL',
    thumbnails: {
      url: 'https://thumbnailurl.com',
      alt: 'some alt text',
    },
    categories: 'about america',
  },
  tableHeaders: [
    { name: 'projectTitle', label: 'PROJECT TITLE' },
    { name: 'updatedAt', label: 'CREATED' },
    { name: 'visibility', label: 'VISIBILITY' },
    { name: 'author', label: 'AUTHOR' },
  ],
  projectTab: 'myProjects',
};

const teamProjectsProps = {
  ...myProjectsProps,
  projectTab: 'teamProjects',
};

const menuItems = [
  { name: 'team', label: 'TEAM' },
  { name: 'categories', label: 'CATEGORIES' },
  { name: 'updatedAt', label: 'MODIFIED DATE' },
];

const MyProjectsComponent = <TableRow { ...myProjectsProps } />;
const TeamProjectsComponent = <TableRow { ...teamProjectsProps } />;

describe( '<TableRow />', () => {
  it( 'renders without crashing', () => {
    const wrapper = shallow( MyProjectsComponent );

    expect( wrapper.exists() ).toEqual( true );
  } );

  it( 'renders the persistent table headers if projectTab is myProjects', () => {
    const wrapper = shallow( MyProjectsComponent );
    const columns = wrapper.find( '.items_table_item' );
    const columnCount = columns.length;
    const expectedColumnCount = myProjectsProps.tableHeaders.length;

    expect( columnCount ).toEqual( expectedColumnCount );
    expect( toJSON( columns ) ).toMatchSnapshot();
  } );

  it( 'renders the persistent table headers if projectTab is teamProjects', () => {
    const wrapper = shallow( TeamProjectsComponent );
    const columns = wrapper.find( '.items_table_item' );
    const columnCount = columns.length;
    const expectedColumnCount = teamProjectsProps.tableHeaders.length;

    expect( columnCount ).toEqual( expectedColumnCount );
    expect( toJSON( columns ) ).toMatchSnapshot();
  } );

  it( 'renders Team, Categories, and Modified Date columns when selected', () => {
    const newProps = {
      ...myProjectsProps,
      tableHeaders: [
        ...myProjectsProps.tableHeaders,
        ...menuItems,
      ],
    };

    const wrapper = shallow( <TableRow { ...newProps } /> );

    const columns = wrapper.find( '.items_table_item' );
    const columnCount = columns.length;
    const expectedColumnCount = newProps.tableHeaders.length;

    expect( columnCount ).toEqual( expectedColumnCount );
    expect( toJSON( columns ) ).toMatchSnapshot();
  } );

  it( 'renders null if myProjectsProps.d is falsy', () => {
    const newProps = { ...myProjectsProps, d: null };
    const wrapper = shallow( <TableRow { ...newProps } /> );

    expect( wrapper.html() ).toEqual( null );
  } );
} );
