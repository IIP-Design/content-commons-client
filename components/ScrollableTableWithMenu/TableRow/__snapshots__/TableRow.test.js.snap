// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TableRow /> renders Team, Categories, and Modified Date columns when selected 1`] = `
Array [
  <TableCell
    as="td"
    className="items_table_item"
    data-header="PROJECT TITLE"
    key="d137_projectTitle"
  >
    <div
      className="primary_col"
    >
      <VideoDetailsPopup
        d={
          Object {
            "author": "Jane Doe",
            "categories": "about america",
            "createdAt": "2019-05-09T18:33:03.368Z",
            "id": "d137",
            "projectTitle": "Test Title 1",
            "status": "PUBLISHED",
            "team": "IIP Video Production",
            "thumbnails": Object {
              "alt": "some alt text",
              "url": "https://thumbnailurl.com",
            },
            "updatedAt": "2019-05-09T18:33:03.368Z",
            "visibility": "INTERNAL",
          }
        }
        header={
          Object {
            "label": "PROJECT TITLE",
            "name": "projectTitle",
          }
        }
      />
    </div>
    <TableMobileDataToggleIcon
      isOpen={false}
      toggleDisplay={[Function]}
    />
  </TableCell>,
  <TableCell
    as="td"
    className="items_table_item"
    data-header="CREATED"
    key="d137_updatedAt"
  >
    <span>
      <div
        className="items_table_mobileHeader"
      >
        CREATED
      </div>
      May 9, 2019
    </span>
    <br />
    <span>
      PUBLISHED
    </span>
  </TableCell>,
  <TableCell
    as="td"
    className="items_table_item"
    data-header="VISIBILITY"
    key="d137_visibility"
  >
    <span>
      <div
        className="items_table_mobileHeader"
      >
        VISIBILITY
      </div>
      INTERNAL
    </span>
    <br />
  </TableCell>,
  <TableCell
    as="td"
    className="items_table_item"
    data-header="AUTHOR"
    key="d137_author"
  >
    <span>
      <div
        className="items_table_mobileHeader"
      >
        AUTHOR
      </div>
      Jane Doe
    </span>
    <br />
  </TableCell>,
  <TableCell
    as="td"
    className="items_table_item"
    data-header="TEAM"
    key="d137_team"
  >
    <span>
      <div
        className="items_table_mobileHeader"
      >
        TEAM
      </div>
      IIP Video Production
    </span>
    <br />
  </TableCell>,
  <TableCell
    as="td"
    className="items_table_item"
    data-header="CATEGORIES"
    key="d137_categories"
  >
    <span>
      <div
        className="items_table_mobileHeader"
      >
        CATEGORIES
      </div>
      about america
    </span>
    <br />
  </TableCell>,
  <TableCell
    as="td"
    className="items_table_item"
    data-header="MODIFIED DATE"
    key="d137_updatedAt"
  >
    <span>
      <div
        className="items_table_mobileHeader"
      >
        MODIFIED DATE
      </div>
      May 9, 2019
    </span>
    <br />
  </TableCell>,
]
`;

exports[`<TableRow /> renders the persistent table headers if projectTab is myProjects 1`] = `
Array [
  <TableCell
    as="td"
    className="items_table_item"
    data-header="PROJECT TITLE"
    key="d137_projectTitle"
  >
    <div
      className="primary_col"
    >
      <VideoDetailsPopup
        d={
          Object {
            "author": "Jane Doe",
            "categories": "about america",
            "createdAt": "2019-05-09T18:33:03.368Z",
            "id": "d137",
            "projectTitle": "Test Title 1",
            "status": "PUBLISHED",
            "team": "IIP Video Production",
            "thumbnails": Object {
              "alt": "some alt text",
              "url": "https://thumbnailurl.com",
            },
            "updatedAt": "2019-05-09T18:33:03.368Z",
            "visibility": "INTERNAL",
          }
        }
        header={
          Object {
            "label": "PROJECT TITLE",
            "name": "projectTitle",
          }
        }
      />
    </div>
    <TableMobileDataToggleIcon
      isOpen={false}
      toggleDisplay={[Function]}
    />
  </TableCell>,
  <TableCell
    as="td"
    className="items_table_item"
    data-header="CREATED"
    key="d137_updatedAt"
  >
    <span>
      <div
        className="items_table_mobileHeader"
      >
        CREATED
      </div>
      May 9, 2019
    </span>
    <br />
    <span>
      PUBLISHED
    </span>
  </TableCell>,
  <TableCell
    as="td"
    className="items_table_item"
    data-header="VISIBILITY"
    key="d137_visibility"
  >
    <span>
      <div
        className="items_table_mobileHeader"
      >
        VISIBILITY
      </div>
      INTERNAL
    </span>
    <br />
  </TableCell>,
  <TableCell
    as="td"
    className="items_table_item"
    data-header="AUTHOR"
    key="d137_author"
  >
    <span>
      <div
        className="items_table_mobileHeader"
      >
        AUTHOR
      </div>
      Jane Doe
    </span>
    <br />
  </TableCell>,
]
`;

exports[`<TableRow /> renders the persistent table headers if projectTab is teamProjects 1`] = `
Array [
  <TableCell
    as="td"
    className="items_table_item"
    data-header="PROJECT TITLE"
    key="d137_projectTitle"
  >
    <div
      className="primary_col"
    >
      <VideoDetailsPopup
        d={
          Object {
            "author": "Jane Doe",
            "categories": "about america",
            "createdAt": "2019-05-09T18:33:03.368Z",
            "id": "d137",
            "projectTitle": "Test Title 1",
            "status": "PUBLISHED",
            "team": "IIP Video Production",
            "thumbnails": Object {
              "alt": "some alt text",
              "url": "https://thumbnailurl.com",
            },
            "updatedAt": "2019-05-09T18:33:03.368Z",
            "visibility": "INTERNAL",
          }
        }
        header={
          Object {
            "label": "PROJECT TITLE",
            "name": "projectTitle",
          }
        }
      />
    </div>
    <TableMobileDataToggleIcon
      isOpen={false}
      toggleDisplay={[Function]}
    />
  </TableCell>,
  <TableCell
    as="td"
    className="items_table_item"
    data-header="CREATED"
    key="d137_updatedAt"
  >
    <span>
      <div
        className="items_table_mobileHeader"
      >
        CREATED
      </div>
      May 9, 2019
    </span>
    <br />
    <span>
      PUBLISHED
    </span>
  </TableCell>,
  <TableCell
    as="td"
    className="items_table_item"
    data-header="VISIBILITY"
    key="d137_visibility"
  >
    <span>
      <div
        className="items_table_mobileHeader"
      >
        VISIBILITY
      </div>
      INTERNAL
    </span>
    <br />
  </TableCell>,
  <TableCell
    as="td"
    className="items_table_item"
    data-header="AUTHOR"
    key="d137_author"
  >
    <span>
      <div
        className="items_table_mobileHeader"
      >
        AUTHOR
      </div>
      Jane Doe
    </span>
    <br />
  </TableCell>,
]
`;
